tracing.go

```diff
@@ -84,53 +94,167 @@ func (s Span) ParentID() uint64 { return s.parentID }
+		return headers
+	}
+	if headers == nil {
+		headers = make(map[string]string)
 	}
+	carrier := tracingHeadersCarrier(headers)
+	if err := span.Tracer().Inject(span.Context(), opentracing.TextMap, carrier); err != nil {
+		// Something had to go seriously wrong for Inject to fail, usually a setup problem.
+		// A good Tracer implementation may also emit a metric.
+		response.log.Error("Failed to inject tracing span: " + err.Error())
+	}
+	return headers
+}
 
-	if rand.Float64() > sampleRate {
-		s.EnableTracing(false)
+// extractInboundSpan attempts to create a new OpenTracing Span for inbound request
+// using only trace IDs stored in the frame's tracing field. It only works if the
+// tracer understand Zipkin-style trace IDs. If such attempt fails, another attempt
+// will be made from the higher level function ExtractInboundSpan() once the
+// application headers are read from the wire.
+func (c *Connection) extractInboundSpan(callReq *callReq) opentracing.Span {
+	spanCtx, err := c.Tracer().Extract(zipkinSpanFormat, &callReq.Tracing)
+	if err != nil {
+		if err != opentracing.ErrUnsupportedFormat && err != opentracing.ErrSpanContextNotFound {
+			c.log.Error("Failed to extract Zipkin-style span: " + err.Error())
+		}
+		return nil
+	}
+	if spanCtx == nil {
+		return nil
 	}
+	operationName := "" // not known at this point, will be set later
+	span := c.Tracer().StartSpan(operationName, ext.RPCServerOption(spanCtx))
+	span.SetTag("as", callReq.Headers[ArgScheme])
+	ext.PeerService.Set(span, callReq.Headers[CallerName])
+	setPeerHostPort(span, c.remotePeerInfo.HostPort)
+	return span
 }
 
-func newSpan(traceID, spanID, parentID uint64, tracingEnabled bool) *Span {
-	flags := byte(0)
-	if tracingEnabled {
-		flags = tracingFlagEnabled
+// ExtractInboundSpan is a higher level version of extractInboundSpan().
+// If the lower-level attempt to create a span from incoming request was
+// successful (e.g. when then Tracer supports Zipkin-style trace IDs),
+// then the application headers are only used to read the Baggage and add
+// it to the existing span. Otherwise, the standard OpenTracing API supported
+// by all tracers is used to deserialize the tracing context from the
+// application headers and start a new server-side span.
+// Once the span is started, it is wrapped in a new Context, which is returned.
+func ExtractInboundSpan(ctx context.Context, call *InboundCall, headers map[string]string, tracer opentracing.Tracer) context.Context {
+	var span = call.Response().span
+	if span != nil {
+		if headers != nil {
+			// extract SpanContext from headers, but do not start another span with it,
+			// just get the baggage and copy to the already created span
+			carrier := tracingHeadersCarrier(headers)
+			if sc, err := tracer.Extract(opentracing.TextMap, carrier); err == nil {
+				sc.ForeachBaggageItem(func(k, v string) bool {
+					span.Context().SetBaggageItem(k, v)
+					return true
+				})
+			}
+			carrier.RemoveTracingKeys()
```

Comment:
@prashantv I've added the namespacing feature for tracing keys by using a custom `tracingHeadersCarrier` which uses a prefix for all tracing keys. The `RemoveTracingKeys()` call restores the application headers to a pre-injection state.

The prefixed and unprefixed keys are cached.